src/Orleans/Streams/AdHoc/IGrainObserver.cs

```diff
@@ -0,0 +1,458 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Threading.Tasks;
+using Orleans.CodeGeneration;
+using Orleans.Concurrency;
+using Orleans.Runtime;
+
+namespace Orleans.Streams.AdHoc
+{
+    /// <summary>
+    /// Represents a resource which can be disposed of asynchronously.
+    /// </summary>
+    public interface IAsyncDisposable
```

Comment:
Suggest IObserverGrain.  Grain observer sounds like we're observing a grain, whereas this is a grain that observes something, hence IObserverGrain.

Similar suggestion for IGrainObservable => IObservableGrain